File: problems-adapt/0238_product-of-the-rest/solution_rolling.cpp
// The judge's expected values are exact big integers, so the products are
// computed with a small base-1e9 bignum. The harness serializes the return
// value via overloaded openoj_json functions, so the bignum provides an
// overload that renders it as a bare integer token.
#include <cstdint>

struct OpenOjBig {
    bool neg = false;
    vector<uint32_t> d; // base 1e9, little-endian; empty = zero
};

static OpenOjBig openojBigOne() { return OpenOjBig{false, {1}}; }

static OpenOjBig openojBigMulSmall(const OpenOjBig &a, long long v) {
    if (v == 0 || a.d.empty()) {
        return OpenOjBig{};
    }
    OpenOjBig r;
    r.neg = a.neg != (v < 0);
    unsigned long long x = v < 0 ? -(unsigned long long)v : (unsigned long long)v;
    unsigned long long carry = 0;
    for (uint32_t digit : a.d) {
        unsigned long long cur = (unsigned long long)digit * x + carry;
        r.d.push_back((uint32_t)(cur % 1000000000ULL));
        carry = cur / 1000000000ULL;
    }
    while (carry) {
        r.d.push_back((uint32_t)(carry % 1000000000ULL));
        carry /= 1000000000ULL;
    }
    return r;
}

static OpenOjBig openojBigMulBig(const OpenOjBig &a, const OpenOjBig &b) {
    if (a.d.empty() || b.d.empty()) {
        return OpenOjBig{};
    }
    OpenOjBig r;
    r.neg = a.neg != b.neg;
    r.d.assign(a.d.size() + b.d.size() + 1, 0);
    for (size_t i = 0; i < a.d.size(); i++) {
        unsigned long long carry = 0;
        size_t j = 0;
        while (j < b.d.size() || carry) {
            unsigned long long cur = (unsigned long long)r.d[i + j] + carry;
            if (j < b.d.size()) {
                cur += (unsigned long long)a.d[i] * b.d[j];
            }
            r.d[i + j] = (uint32_t)(cur % 1000000000ULL);
            carry = cur / 1000000000ULL;
            j++;
        }
    }
    while (!r.d.empty() && r.d.back() == 0) {
        r.d.pop_back();
    }
    return r;
}

static string openojBigText(const OpenOjBig &b) {
    if (b.d.empty()) {
        return "0";
    }
    string s = b.neg ? "-" : "";
    s += to_string(b.d.back());
    for (size_t i = b.d.size() - 1; i-- > 0;) {
        char buf[16];
        snprintf(buf, sizeof(buf), "%09u", b.d[i]);
        s += buf;
    }
    return s;
}

static string openoj_json(const OpenOjBig &value) { return openojBigText(value); }

class Solution {
  public:
    vector<OpenOjBig> productOfRest(vector<int> &nums) {
        // The product except nums[i] factors as (product of everything
        // before i) x (product of everything after i), both computable as
        // running products — no division, which zeros would break anyway.
        int n = nums.size();
        vector<OpenOjBig> answer(n, openojBigOne());
        // First sweep stores the running left product BEFORE folding nums[i]
        // in, so answer[i] ends up holding exactly the prefix preceding i.
        OpenOjBig left = openojBigOne();
        for (int i = 0; i < n; i++) {
            answer[i] = left;
            left = openojBigMulSmall(left, nums[i]);
        }
        // Second sweep from the right: its running product likewise lags one
        // position behind, then absorbs nums[i]. Each cell becomes
        // prefix x suffix.
        OpenOjBig right = openojBigOne();
        for (int i = n - 1; i >= 0; i--) {
            answer[i] = openojBigMulBig(answer[i], right);
            right = openojBigMulSmall(right, nums[i]);
        }
        // Zeros need no special casing: a lone zero zeroes every cell but its
        // own, and multiple zeros zero everything — all automatic.
        return answer;
    }
};

File: problems-bettercode/0201-0300/0238_product-of-array-except-self/solution_rolling.go
// The judge's expected values are exact big integers, so the products are
// computed with math/big-style arithmetic (implemented directly on digit
// slices). openojRawInt serializes as a bare JSON number token via its
// MarshalJSON method; the harness marshals the return value with encoding/json.
type openojRawInt string

func (v openojRawInt) MarshalJSON() ([]byte, error) { return []byte(v), nil }

// openojBig is a signed big integer stored base 1e9, little-endian.
// An empty digit slice represents zero.
type openojBig struct {
	neg bool
	d   []uint32
}

func openojBigMulSmall(a openojBig, v int64) openojBig {
	if v == 0 || len(a.d) == 0 {
		return openojBig{}
	}
	neg := a.neg != (v < 0)
	x := uint64(v)
	if v < 0 {
		x = uint64(-v)
	}
	d := make([]uint32, 0, len(a.d)+2)
	var carry uint64
	for _, digit := range a.d {
		cur := uint64(digit)*x + carry
		d = append(d, uint32(cur%1000000000))
		carry = cur / 1000000000
	}
	for carry > 0 {
		d = append(d, uint32(carry%1000000000))
		carry /= 1000000000
	}
	return openojBig{neg: neg, d: d}
}

func openojBigMulBig(a, b openojBig) openojBig {
	if len(a.d) == 0 || len(b.d) == 0 {
		return openojBig{}
	}
	d := make([]uint32, len(a.d)+len(b.d)+1)
	for i := 0; i < len(a.d); i++ {
		var carry uint64
		j := 0
		for j < len(b.d) || carry > 0 {
			cur := uint64(d[i+j]) + carry
			if j < len(b.d) {
				cur += uint64(a.d[i]) * uint64(b.d[j])
			}
			d[i+j] = uint32(cur % 1000000000)
			carry = cur / 1000000000
			j++
		}
	}
	for len(d) > 0 && d[len(d)-1] == 0 {
		d = d[:len(d)-1]
	}
	return openojBig{neg: a.neg != b.neg, d: d}
}

func (b openojBig) String() string {
	if len(b.d) == 0 {
		return "0"
	}
	s := ""
	if b.neg {
		s = "-"
	}
	s += fmt.Sprintf("%d", b.d[len(b.d)-1])
	for i := len(b.d) - 2; i >= 0; i-- {
		s += fmt.Sprintf("%09d", b.d[i])
	}
	return s
}

func productExceptSelf(nums []int) []any {
	// The product except nums[i] factors as (product of everything
	// before i) x (product of everything after i), both computable as
	// running products — no division, which zeros would break anyway.
	n := len(nums)
	answer := make([]openojBig, n)
	// First sweep stores the running left product BEFORE folding nums[i] in,
	// so answer[i] ends up holding exactly the prefix preceding i.
	left := openojBig{d: []uint32{1}}
	for i := 0; i < n; i++ {
		answer[i] = left
		left = openojBigMulSmall(left, int64(nums[i]))
	}
	// Second sweep from the right: its running product likewise lags one
	// position behind, then absorbs nums[i]. Each cell becomes
	// prefix x suffix.
	right := openojBig{d: []uint32{1}}
	for i := n - 1; i >= 0; i-- {
		answer[i] = openojBigMulBig(answer[i], right)
		right = openojBigMulSmall(right, int64(nums[i]))
	}
	// Zeros need no special casing: a lone zero zeroes every cell but its
	// own, and multiple zeros zero everything — all automatic.
	out := make([]any, n)
	for i := range answer {
		out[i] = openojRawInt(answer[i].String())
	}
	return out
}
